cmake/modules/FindSYMENGINE.cmake

```diff
@@ -0,0 +1,129 @@
+##
+## Copyright (C) 2019 by the deal.II authors
+##
+## This file is part of the deal.II library.
+##
+## The deal.II library is free software; you can use it, redistribute
+## it, and/or modify it under the terms of the GNU Lesser General
+## Public License as published by the Free Software Foundation; either
+## version 2.1 of the License, or (at your option) any later version.
+## The full text of the license can be found in the file LICENSE at
+## the top level of the deal.II distribution.
+##
+## ---------------------------------------------------------------------
+
+#
+# - Try to find SymEngine
+#
+# This module exports
+#
+#   SYMENGINE_INCLUDE_DIR
+#   SYMENGINE_LIBRARY
+#   SYMENGINE_WITH_LLVM
+#
+
+SET(SYMENGINE_DIR "" CACHE PATH "An optional hint to a SymEngine installation")
+SET_IF_EMPTY(SYMENGINE_DIR "$ENV{SYMENGINE_DIR}")
+
+#
+# SymEngine overwrites the CMake module path, so we save
+# and restore it after this library is found and configured.
+#
+SET (DEAL_II_CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH})
+
+#
+# Include the SymEngine:
+#
+FIND_PACKAGE(SymEngine ${SYMENGINE_MINIMUM_REQUIRED_VERSION}
+  CONFIG QUIET
+  HINTS ${SYMENGINE_DIR}
+  PATH_SUFFIXES lib/cmake/symengine
+  NO_SYSTEM_ENVIRONMENT_PATH
+  )
+
+#
+# Reset the CMake module path
+#
+SET (CMAKE_MODULE_PATH ${DEAL_II_CMAKE_MODULE_PATH})
+
+
+#
+# Look for symengine_config.h - we'll query it to determine supported features:
+#
+IF(SymEngine_FOUND)
+  DEAL_II_FIND_FILE(SYMENGINE_SETTINGS_H symengine_config.h
+    HINTS ${SYMENGINE_INCLUDE_DIRS}
+    PATH_SUFFIXES symengine
+    NO_DEFAULT_PATH
+    NO_CMAKE_ENVIRONMENT_PATH
+    NO_CMAKE_PATH
+    NO_SYSTEM_ENVIRONMENT_PATH
+    NO_CMAKE_SYSTEM_PATH
+    NO_CMAKE_FIND_ROOT_PATH
+    )
+ENDIF()
+
+#
+# Version check
+#
+IF(EXISTS ${SYMENGINE_SETTINGS_H})
+
+  FILE(STRINGS "${SYMENGINE_SETTINGS_H}" SYMENGINE_VERSION_MAJOR_STRING
+    REGEX "#define.*SYMENGINE_MAJOR_VERSION")
+  STRING(REGEX REPLACE "^.*SYMENGINE_MAJOR_VERSION.*([0-9]+).*" "\\1"
+    SYMENGINE_VERSION_MAJOR "${SYMENGINE_VERSION_MAJOR_STRING}"
+    )
+  FILE(STRINGS "${SYMENGINE_SETTINGS_H}" SYMENGINE_VERSION_MINOR_STRING
+    REGEX "#define.*SYMENGINE_MINOR_VERSION")
+  STRING(REGEX REPLACE "^.*SYMENGINE_MINOR_VERSION.*([0-9]+).*" "\\1"
+    SYMENGINE_VERSION_MINOR "${SYMENGINE_VERSION_MINOR_STRING}"
+    )
+  FILE(STRINGS "${SYMENGINE_SETTINGS_H}" SYMENGINE_VERSION_PATCH_STRING
+    REGEX "#define.*SYMENGINE_PATCH_VERSION")
+  STRING(REGEX REPLACE "^.*SYMENGINE_PATCH_VERSION.*([0-9]+).*" "\\1"
+    SYMENGINE_VERSION_PATCH "${SYMENGINE_VERSION_PATCH_STRING}"
+    )
+  SET(SYMENGINE_VERSION
+    "${SYMENGINE_VERSION_MAJOR}.${SYMENGINE_VERSION_MINOR}.${SYMENGINE_VERSION_PATCH}"
+    )
+ENDIF()
+
+#
+# Feature checks
+#
+
+MACRO(_symengine_feature_check _var _regex)
+  IF(EXISTS ${SYMENGINE_SETTINGS_H})
+    FILE(STRINGS "${SYMENGINE_SETTINGS_H}" SYMENGINE_${_var}_STRING
+      REGEX "${_regex}")
+    IF("${SYMENGINE_${_var}_STRING}" STREQUAL "")
```

Comment:
This can be simplified to,
```
IF("${SYMENGINE_${_var}_INCLUDE_DIRS}" STREQUAL "")
```
and there's no need to check the header file.

Reply:
Ok, great. If you wouldn't mind, please ping me on that PR so that I know to update this.

Reply:
See https://github.com/symengine/symengine/pull/1528.

`HAVE_SYMENGINE_LLVM` will be `True` or empty depending on symengine is compiled with LLVM or not

Reply:
I guess that this still does not completely satisfy what we're wanting to do here. We actually extract these features to automatically enable/disable tests in CMake, not to be used as `#define`s in our code-base itself (for this ). So if I call one of our unit tests blessed outputs `basic_04.with_symengine_with_llvm=false.output`, then this will be the result checked when SymEngine is not built with LLVM. When it is built with LLVM enabled, `basic_04.with_symengine_with_llvm=true.output` would be used for comparison. For this to be possible, the CMake variable must match the template `<LIB_NAME>_WITH_<FEATURE_NAME>`. So we could do the same for `HAVE_SYMENGINE_RESERVE`, if we wanted to.

By the way, is the mix of `WITH_SYMENGINE_<>` and `HAVE_SYMENGINE_<>` in `symengine_config.h` intended?

Reply:
You could do,
```cmake
if ("${HAVE_SYMENGINE_LLVM}" STREQUAL "True")
   set (SYMENGINE_WITH_LLVM True)
else ()
   set (SYMENGINE_WITH_LLVM False)
endif()
```

Reply:
Are all features accessible like this, or only packages? I.e. could we do 
```cmake
if ("${HAVE_SYMENGINE_NOEXCEPT}" STREQUAL "True")
   set (SYMENGINE_WITH_NOEXCEPT True)
else ()
   set (SYMENGINE_WITH_NOEXCEPT False)
endif()
```
? Sorry for my naivety, my CMake feels a bit rusty right now.

Reply:
Only packages for now, but I can easily add all `HAVE_*` features

Reply:
@tamiko Do you have an opinion on this? This is how we do feature detection for almost all other packages, and its been robust so far. Personally I don't think that its worth fussing over, but since @tamiko is our build system maintainer he's more qualified to weigh up the pro's and con's of both approaches.